### ISSUE TITLE
VIP: Hash-based storage slots

### ISSUE THREAD

**Issue by fubuloubu** [ethereum/vyper#1733]
## Simple Summary
Change the initial storage slot index of a variable from a compilation-specific ordering to a hash-based index based on variable name.

## Motivation
We currently do not do any sort of tightly packing of storage slots in Vyper contracts. The original specification of Vyper's ordering-based indexing of variable storage slots is [adapted from Solidity](https://solidity.readthedocs.io/en/v0.5.13/miscellaneous.html#layout-of-state-variables-in-storage). Unfortunately, due to the compiler making decisions about the ordering of storage slots, there is [chaotic behavior](https://blog.trailofbits.com/2018/09/05/contract-upgrade-anti-patterns/) of storage slot indexes between versions of different contracts, leading to issues when using the Proxy pattern of upgradeable contracts between multiple versions of a Solidity contract.

By creating a predictable storage slot index strategy (based on a collision-resistant, cryptographic hash function such as `keccak256`), external tools can regenerate the entire symbol table of a Vyper contract from nothing else than analyzing source code directly. More importantly, the Proxy pattern would be more resilient with hash-based indexes because the variable name now determines it's location, and not it's ordering within a contract (or inherited system of contracts, a la Solidity)

## Specification
The change asked for here is to change the storage slot index function `index(v)` from a stateful function which auto-increments the storage slot based on variable size of the variables laid out before it, to a stateless function that computes it via the hash of the variables name e.g. `index(v) := keccak(v.name)`. This does not change how structs or arrays are currently laid out (as specified in #769), only the base index for storage of a particular type.

## Backwards Compatibility
This VIP will change how storage slots are calculated, and will break any existing tools that rely on the assumption that storage slots are auto-incremented based on the contract's definitions. However, tools should appreciate that storage slots are now predictably obtained from their names only, without relying on internal compiler information such as how the compiler makes a determination of storage slots based on complex inheritance and compression rules.

## Dependencies
No dependencies. Related to #769.

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/)

**Comment by fubuloubu** [comment]
As a side note, with this VIP implemented it would be possible to implement the Proxy pattern by leveraging Vyper contracts as a template via `DELEGATECALL` and delegating to the specific deployed version you intend, which could be managed on-chain via ethPM registration (or similar). The Proxy itself may or may not be developed with Solidity, that is up to the implementer of the specific framework which decides to implement this pattern for users using the dependable storage slot index provided by this VIP.

**Comment by Pet3ris** [comment]
@fubuloubu this is great! Another benefit is that it helps with implementing tools like debuggers - they wouldn't have to be coupled to a compiler version dependent re: https://github.com/ethereum/vyper/issues/1629.

**Comment by fubuloubu** [comment]
One potential issue is that someone could maliciously define a variable name such that there is a collision:
```python
baz: map(bytes32, Baz)
f2d05ec5c5729fb559780c70a93ca7b4ee2ca37f64e62fa31046b324f60d9447: map(bytes32, Baz)
```

The solution to this could be to throw on variables that look like a hash (because no one should be doing that), or any variables that have the same hash as another previously defined (which should be unlikely given second preimage resistance of keccak)

**Comment by fubuloubu** [comment]
Another assumption is that string types are disallowed as keys. If that assumption were relaxed, the following would be possible:
```python
baz: map(string, Baz)
f2d05ec5c5729fb559780c70a93ca7b4ee2ca37f64e62fa31046b324f60d9447baz: Baz
```

As the two would have the same hash preimage e.g. `keccak(keccak("baz") . "baz") == keccak(f2d0....9447baz)`.

The solution to this would be a domain separator to ensure that such variable names are not possible to construct e.g. `index(v) := keccak("#" . v.name)`

**Comment by fubuloubu** [comment]
Related: https://en.wikipedia.org/wiki/De_Bruijn_index

**Comment by fubuloubu** [comment]
Note: This would increase deployment costs because you would have to do a full `PUSH32` to load the hash index onto the stack for normal variables (same gas cost as `PUSH1`, but more data to push... hence more data in deployment transaction)

**Comment by fubuloubu** [comment]
Note to potential implementers of proxying via this method:

It is wise to protect the parent contract's private storage variables from being overwrite-able by the children's logic by utilizing a seed or domain separator on the parent's storage to ensure it is not possible.

**Comment by thegostep** [comment]
Would love to see this implemented. This makes upgradeability significantly safer. It also makes developing against mainnet so much easier because it becomes trivial to pull a snapshot of the state of a smart contract into a local environment like ganache / buidler.

**Comment by fubuloubu** [comment]
@thegostep Excellent point about pulling state changes!

**Comment by charles-cooper** [comment]
> One potential issue is that someone could maliciously define a variable name such that there is a collision:
> 
> ```python
> baz: map(bytes32, Baz)
> f2d05....9447: map(bytes32, Baz)
> ```
> 
> The solution to this could be to throw on variables that look like a hash (because no one should be doing that), or any variables that have the same hash as another previously defined (which should be unlikely given second preimage resistance of keccak)

wouldn't somebody have to find a keccak collision (which is currently not known to be possible)?

**Comment by charles-cooper** [comment]
My biggest concern about this is that locations become compiler-dependent when we implement https://github.com/vyperlang/vyper/issues/1556. So with something like
```
bools: bool[128]
```
`bools` is presumably always going to start in the same place, but depending on how the compiler packs arrays `bools[1]` could be in a different place.

One way of dealing with this is to mangle the packing strategy and type into the variable name (which is a kind of version namespacing). So, a regular (i.e. unpacked) bool array `bools: bool[128]` might be stored at `keccak('bools: bool[128]')`, whereas say we pack every bool into a single bit, then `bools: bool[128]` might be stored at `keccak('bools: bool#BIT_PACKED[128]')`. If we decide to pack every bool into a byte, then `bools: bool[128]` might be stored at `keccak('bools: bool#BYTE_PACKED[128])`. This kind of makes sense, because even though they might represent the same thing semantically, a packed bool array is literally represented differently in storage than an unpacked bool array. And this is visible to clients as well - if you call `eth_getStorage` (which returns an entire word) at the array location, you will get a different answer depending on the packing strategy.

So maybe this just seems like a good idea because I'm up in the middle of the night but it seems putting the type in the preimage is also kind of neat because it provides some kind of type safety to variable slots (e.g. `xs: uint256` and `xs: int128` are stored in different locations). It prevents contract writers from ex post facto unsafely coercing a storage slot to another type (although this might be considered a shortcoming, if that is something that the contract writer really, really wants to do).

**Comment by fubuloubu** [comment]
> wouldn't somebody have to find a keccak collision (which is currently not known to be possible)?

No, you could literally just call the map by it's preimage, and it would have the same slot index by definition... although, thinking about it now... that wouldn't really be practical given we encode to bytes first because you'd have to make up a garbage name that wouldn't be valid. I think @smarx was thinking mostly if it was string-encoded it would be easy to do this.

**Comment by fubuloubu** [comment]
> My biggest concern about this is that locations become compiler-dependent when we implement #1556.

Yes, but that would only be a difference between versions (i.e. once you upgraded to that version of vyper, it should always do the same thing)

I think it's a really good idea though to separate the domain a bit here via version of this proposal, so we have space to say "okay, we need to change what `v1` said this encoding was, use `v2`". At that point, we would also need to introduce a syntax for inter-operating with older versions in code, but it could be done (somehow, maybe decorators?)

**Comment by fubuloubu** [comment]
> (although this might be considered a shortcoming, if that is something that the contract writer really, really wants to do)

Wanting to add another member to a struct is a prime example of where this would have adverse effects

**Comment by smarx** [comment]
No, I wasn't thinking about hexadecimal encoding (but I did use that shorthand as a way to explain the issue). But rather than worry about the odds a hash will consist solely of valid characters for a variable identifier, it's probably better to just avoid the possibility of an _input_ collision altogether. It's easy to do via a prefix or separator that can't occur in identifiers, like "#" or "\0".

**Comment by fubuloubu** [comment]
> One way of dealing with this is to mangle the packing strategy

This isn't necessarily the worst idea though, we could define a decorator e.g. `bits: packed(bool[128])` that does this same thing, AND makes it more clear in code that you are packing i.e. it becomes an opt-in strategy

**Comment by fubuloubu** [comment]
> No, I wasn't thinking about hexadecimal encoding (but I did use that shorthand as a way to explain the issue). But rather than worry about the odds a hash will consist solely of valid characters for a variable identifier, it's probably better to just avoid the possibility of an _input_ collision altogether. It's easy to do via a prefix or separator that can't occur in identifiers, like "#" or "\0".

Yes but the string version of the hash would not be `abc123...`, it would be `Iz@n;1yBzc&KnvY?D[F/...` which would not be a valid Vyper identifier

**Comment by fubuloubu** [comment]
Finding an _actual_ pre-image collision would be difficult because of keccak's 2nd preimage collision resistance, and it would most likely not be a valid version identifier anyways, so I don't think it's really a practical concern.

**Comment by smarx** [comment]
> Yes but the string version of the hash would not be abc123..., it would be Iz@n;1yBzc&KnvY?D[F/... which would not be a valid Vyper identifier

_Most_ hashes would not be valid Vyper identifiers. Some would. I don't really care to figure out how many because this problem is trivial to solve.

**Comment by fubuloubu** [comment]
I still think the likelihood is astronomically small, but regardless it's definitely a good idea to have a domain separator anyways, and I think @charles-cooper's suggestion of versioning of this mechanism being incorporated into the domain separation is what we need.

The revised suggestion would be something like:
```
index(v) := keccak('{index-version}#' . v.name [. '#{packing-version}'])
```

**Comment by jacqueswww** [comment]
Meeting notes:
- Provisionally approved; 
- The recommended approach would be to implement something similar to memoryallocator class that would allow easy retrieval of a storage slot first.
- Then we can implement alternate strategies like this one.

**Comment by fubuloubu** [comment]
Other ideas: have the storage allocation strategy be configurable via compiler flag or external API

**Comment by fubuloubu** [comment]
Note: Internal variables must have their own namespace
